src/messenger.rs

```diff
@@ -922,6 +932,122 @@ mod tests {
         assert_eq!(actual, resp);
     }
 
+    #[tokio::test]
+    async fn test_cancel_request() {
```

Comment:
I think this test would be easier to follow if the loops were broken up into separate functions, but it's fine as is also